File: pingdartdb-node/src/QueryBuilder.js
class QueryBuilder {
    constructor(tableName, sequelize, dbType) {
        this.tableName = tableName;
        this.sequelize = sequelize;
        this.dbType = dbType;
        this.q = this.dbType === 'postgres' ? '"' : '`';
    }

    /**
     * Helper to build dynamic WHERE clauses with advanced operators
     */
    _buildWhereClause(conditions, replacements, prefix = 'cond_') {
        const clauses = [];
        if (!conditions || typeof conditions !== 'object') return clauses;

        for (const [key, value] of Object.entries(conditions)) {
            const paramKey = `${prefix}${key}`;
            
            if (Array.isArray(value)) {
                if (value.length > 0) {
                    const placeholders = value.map((_, i) => `:${paramKey}_${i}`).join(', ');
                    clauses.push(`${this.q}${key}${this.q} IN (${placeholders})`);
                    value.forEach((v, i) => { replacements[`${paramKey}_${i}`] = v; });
                } else {
                    clauses.push('1=0');
                }
            } else if (typeof value === 'object' && value !== null) {
                if ((value.start !== undefined && value.end !== undefined) || 
                    (value.min !== undefined && value.max !== undefined)) {
                    const startValue = value.start !== undefined ? value.start : value.min;
                    const endValue = value.end !== undefined ? value.end : value.max;
                    clauses.push(`${this.q}${key}${this.q} BETWEEN :${paramKey}_start AND :${paramKey}_end`);
                    replacements[`${paramKey}_start`] = startValue;
                    replacements[`${paramKey}_end`] = endValue;
                } else if (value instanceof Date) {
                    clauses.push(`${this.q}${key}${this.q} = :${paramKey}`);
                    replacements[paramKey] = value;
                }
            } else if (typeof value === 'string') {
                if (value.startsWith('!')) {
                    clauses.push(`${this.q}${key}${this.q} != :${paramKey}`);
                    replacements[paramKey] = value.substring(1);
                } else if (value.startsWith('>=') || value.startsWith('=>')) {
                    clauses.push(`${this.q}${key}${this.q} >= :${paramKey}`);
                    replacements[paramKey] = value.substring(2);
                } else if (value.startsWith('<=') || value.startsWith('=<')) {
                    clauses.push(`${this.q}${key}${this.q} <= :${paramKey}`);
                    replacements[paramKey] = value.substring(2);
                } else if (value.startsWith('>')) {
                    clauses.push(`${this.q}${key}${this.q} > :${paramKey}`);
                    replacements[paramKey] = value.substring(1);
                } else if (value.startsWith('<')) {
                    clauses.push(`${this.q}${key}${this.q} < :${paramKey}`);
                    replacements[paramKey] = value.substring(1);
                } else {
                    clauses.push(`${this.q}${key}${this.q} = :${paramKey}`);
                    replacements[paramKey] = value;
                }
            } else {
                clauses.push(`${this.q}${key}${this.q} = :${paramKey}`);
                replacements[paramKey] = value;
            }
        }
        return clauses;
    }

    /**
     * Helper to process margedata (Batch joins) recursively
     */
    async _processMergedataBatch(rows, margedataItem, search = {}) {
        const { target_table, target_column, target_value, target_label, search_fields = [], margedata: nestedMargedata, range } = margedataItem;
        if (!target_table || !target_column || !target_value || !target_label) return;

        try {
            const parentIds = [...new Set(rows.map(row => row[target_value]).filter(Boolean))];
            if (parentIds.length === 0) {
                rows.forEach(row => { row[target_label] = []; });
                return;
            }

            const conditions = [`${this.q}${target_column}${this.q} IN (:parentIds)`];
            const replacements = { parentIds };

            for (const field of search_fields) {
                for (const [searchKey, fullColumn] of Object.entries(field)) {
                    const parts = fullColumn.split(".");
                    const column = parts.length > 1 ? parts[1] : parts[0];
                    
                    if (search[searchKey]) {
                        const likeOp = this.dbType === 'postgres' ? 'ILIKE' : 'LIKE';
                        conditions.push(`${this.q}${column}${this.q} ${likeOp} :${searchKey}`);
                        replacements[searchKey] = `%${search[searchKey]}%`;
                    }
                }
            }

            // Geolocation Range Support within Margedata
            if (range && typeof range === 'object' && range.latitude && range.longitude && range.radius && range.target_latitude && range.target_longitude) {
                const { latitude, longitude, radius, target_latitude, target_longitude } = range;
                const earthRadius = 6371; // KM
                
                conditions.push(`(
                    ${earthRadius} * ACOS(
                        COS(RADIANS(:_batchLat)) *
                        COS(RADIANS(${this.q}${target_latitude}${this.q})) *
                        COS(RADIANS(${this.q}${target_longitude}${this.q}) - RADIANS(:_batchLon)) +
                        SIN(RADIANS(:_batchLat)) *
                        SIN(RADIANS(${this.q}${target_latitude}${this.q}))
                    )
                ) <= :_batchRad`);

                replacements._batchLat = latitude;
                replacements._batchLon = longitude;
                replacements._batchRad = radius;
            }

            const batchQuery = `SELECT * FROM ${this.q}${target_table}${this.q} WHERE ${conditions.join(' AND ')}`;
            
            const allMatches = await this.sequelize.query(batchQuery, {
                replacements,
                type: this.sequelize.QueryTypes.SELECT,
            });

            const groupedMatches = {};
            allMatches.forEach(match => {
                const parentId = match[target_column];
                if (!groupedMatches[parentId]) groupedMatches[parentId] = [];
                groupedMatches[parentId].push(match);
            });

            rows.forEach(row => {
                const parentId = row[target_value];
                row[target_label] = groupedMatches[parentId] || [];
            });

            if (Array.isArray(nestedMargedata) && nestedMargedata.length > 0 && allMatches.length > 0) {
                for (const childItem of nestedMargedata) {
                    await this._processMergedataBatch(allMatches, childItem, search);
                }
            }
        } catch (err) {
            console.error(`Error in batch margedata (${target_label}):`, err.message);
            rows.forEach(row => { row[target_label] = []; });
        }
    }

    /**
     * Read records with dynamic conditions, operators, geolocation, and margedata joins
     * @param {Object} options - { conditions, search, orderBy, pagination, range, margedata }
     */
    async read(options = {}) {
        const { conditions, search, orderBy, pagination, range, margedata } = options;
        let query = `SELECT * FROM ${this.q}${this.tableName}${this.q} WHERE 1=1`;
        const replacements = {};

        // Advanced Conditions
        const conditionClauses = this._buildWhereClause(conditions, replacements, 'r_');
        if (conditionClauses.length > 0) {
            query += ` AND ${conditionClauses.join(' AND ')}`;
        }

        // Search (LIKE / ILIKE)
        if (search && typeof search === 'object') {
            const searchClauses = [];
            for (const [key, value] of Object.entries(search)) {
                if (value) {
                    const likeOp = this.dbType === 'postgres' ? 'ILIKE' : 'LIKE';
                    searchClauses.push(`${this.q}${key}${this.q} ${likeOp} :s_${key}`);
                    replacements[`s_${key}`] = `%${value}%`;
                }
            }
            if (searchClauses.length > 0) {
                query += ` AND (${searchClauses.join(' OR ')})`;
            }
        }

        // Geolocation Range (Supports Case 1: Same Table & Case 2: Join Table)
        if (range && typeof range === 'object' && range.latitude && range.longitude && range.radius && range.target_latitude && range.target_longitude) {
            const { latitude, longitude, radius, target_latitude, target_longitude, table_name, join_column = 'user_id', main_join_column = 'id' } = range;
            const earthRadius = 6371; // KM

            if (!table_name || table_name === this.tableName) {
                // Case 1: Same table
                query += ` AND (
                    ${earthRadius} * ACOS(
                        COS(RADIANS(:_rLat)) *
                        COS(RADIANS(${this.q}${target_latitude}${this.q})) *
                        COS(RADIANS(${this.q}${target_longitude}${this.q}) - RADIANS(:_rLon)) +
                        SIN(RADIANS(:_rLat)) *
                        SIN(RADIANS(${this.q}${target_latitude}${this.q}))
                    )
                ) <= :_rRad`;
            } else {
                // Case 2: Join with location table
                const alias = 'rt';
                query += ` AND EXISTS (
                    SELECT 1 FROM ${this.q}${table_name}${this.q} AS ${alias}
                    WHERE ${alias}.${this.q}${join_column}${this.q} = ${this.q}${this.tableName}${this.q}.${this.q}${main_join_column}${this.q}
                    AND (
                        ${earthRadius} * ACOS(
                            COS(RADIANS(:_rLat)) *
                            COS(RADIANS(${alias}.${this.q}${target_latitude}${this.q})) *
                            COS(RADIANS(${alias}.${this.q}${target_longitude}${this.q}) - RADIANS(:_rLon)) +
                            SIN(RADIANS(:_rLat)) *
                            SIN(RADIANS(${alias}.${this.q}${target_latitude}${this.q}))
                        )
                    ) <= :_rRad
                )`;
            }
            
            replacements._rLat = latitude;
            replacements._rLon = longitude;
            replacements._rRad = radius;
        }

        if (orderBy) query += ` ORDER BY ${orderBy}`;

        // Get total count BEFORE pagination
        const countQuery = query.replace('SELECT *', 'SELECT COUNT(*) as total');
        const countResult = await this.sequelize.query(countQuery, {
            replacements,
            type: this.sequelize.QueryTypes.SELECT,
            raw: true
        });
        const totalCount = parseInt(countResult[0]?.total || 0, 10);

        // Pagination setup
        let totalPages = 0;
        if (pagination && pagination.page && pagination.limit) {
            const page = Math.max(1, parseInt(pagination.page, 10));
            const limit = parseInt(pagination.limit, 10);
            const offset = (page - 1) * limit;
            query += ` LIMIT ${limit} OFFSET ${offset}`;
            totalPages = Math.ceil(totalCount / limit);
        }

        // Execute Paginated Query
        const results = await this.sequelize.query(query, {
            replacements,
            type: this.sequelize.QueryTypes.SELECT,
            raw: true
        });

        // Margedata (Batch Relation Fetches)
        if (margedata && Array.isArray(margedata) && margedata.length > 0 && results.length > 0) {
            for (const mdItem of margedata) {
                await this._processMergedataBatch(results, mdItem, search);
            }
        }

        return { success: true, data: results, totalCount, totalPages };
    }

    /**
     * Insert one or multiple records (Bulk Insert) with optional duplicate check
     * @param {Object} [options] - Optional options object, e.g., { conditions: { email: '...' } }. If provided, checks if a matching record already exists before inserting.
     * @param {Object|Array} data - Record object or array of objects to insert
     */
    async insert(arg1, arg2) {
        let data, conditions = null;

        if (arg2 !== undefined) {
            // Signature: insert(options, data)
            data = arg2;
            if (arg1 && typeof arg1 === 'object' && arg1.conditions) {
                conditions = arg1.conditions;
            }
        } else {
            // Signature: insert(data)
            data = arg1;
        }

        if (!data) throw new Error("Data object cannot be empty");

        const isBulk = Array.isArray(data);
        const records = isBulk ? data : [data];

        if (records.length === 0) throw new Error("Data array cannot be empty");

        // --- Duplicate Check ---
        if (conditions && typeof conditions === 'object' && Object.keys(conditions).length > 0) {
            if (!isBulk) {
                // Single insert: check if a record matching `conditions` already exists
                const checkReplacements = {};
                const conditionClauses = this._buildWhereClause(conditions, checkReplacements, 'ins_chk_');
                if (conditionClauses.length > 0) {
                    const checkQuery = `SELECT COUNT(*) as total FROM ${this.q}${this.tableName}${this.q} WHERE ${conditionClauses.join(' AND ')}`;
                    const checkResult = await this.sequelize.query(checkQuery, {
                        replacements: checkReplacements,
                        type: this.sequelize.QueryTypes.SELECT,
                        raw: true
                    });
                    const existingCount = parseInt(checkResult[0]?.total || 0, 10);
                    if (existingCount > 0) {
                        return {
                            success: false,
                            exists: true,
                            message: 'Record already exists matching the given conditions. Insert skipped.'
                        };
                    }
                }
            } else {
                // Bulk insert: filter out records that already exist based on the condition keys
                const conditionKeys = Object.keys(conditions);
                const toInsert = [];
                const skipped = [];

                for (const record of records) {
                    // Build per-record conditions from the condition keys using values from the record
                    const perRecordConditions = {};
                    for (const key of conditionKeys) {
                        if (record[key] !== undefined) {
                            perRecordConditions[key] = record[key];
                        } else {
                            // Use the static condition value from the conditions object
                            perRecordConditions[key] = conditions[key];
                        }
                    }
                    const checkReplacements = {};
                    const conditionClauses = this._buildWhereClause(perRecordConditions, checkReplacements, `ins_bulk_chk_`);
                    if (conditionClauses.length > 0) {
                        const checkQuery = `SELECT COUNT(*) as total FROM ${this.q}${this.tableName}${this.q} WHERE ${conditionClauses.join(' AND ')}`;
                        const checkResult = await this.sequelize.query(checkQuery, {
                            replacements: checkReplacements,
                            type: this.sequelize.QueryTypes.SELECT,
                            raw: true
                        });
                        const existingCount = parseInt(checkResult[0]?.total || 0, 10);
                        if (existingCount > 0) {
                            skipped.push(record);
                        } else {
                            toInsert.push(record);
                        }
                    } else {
                        toInsert.push(record);
                    }
                }

                if (toInsert.length === 0) {
                    return {
                        success: false,
                        exists: true,
                        message: `All ${records.length} records already exist. Insert skipped.`,
                        skipped: skipped.length
                    };
                }

                // Only insert non-duplicate records
                const keys = Object.keys(toInsert[0]);
                const columns = keys.map(k => `${this.q}${k}${this.q}`).join(', ');
                const replacements = [];
                const valueSets = [];
                for (const record of toInsert) {
                    const placeholders = [];
                    for (const key of keys) {
                        placeholders.push('?');
                        replacements.push(record[key]);
                    }
                    valueSets.push(`(${placeholders.join(', ')})`);
                }
                let query = `INSERT INTO ${this.q}${this.tableName}${this.q} (${columns}) VALUES ${valueSets.join(', ')}`;
                if (this.dbType === 'postgres') query += ' RETURNING *';
                const [results] = await this.sequelize.query(query, {
                    replacements,
                    type: this.sequelize.QueryTypes.INSERT
                });
                return {
                    success: true,
                    message: `${toInsert.length} record(s) inserted. ${skipped.length} duplicate(s) skipped.`,
                    inserted: toInsert.length,
                    skipped: skipped.length,
                    insertedRows: results
                };
            }
        }

        // --- Standard Insert (no conditions check) ---
        const keys = Object.keys(records[0]);
        const columns = keys.map(k => `${this.q}${k}${this.q}`).join(', ');

        const replacements = [];
        const valueSets = [];

        for (const record of records) {
            const placeholders = [];
            for (const key of keys) {
                placeholders.push('?');
                replacements.push(record[key]);
            }
            valueSets.push(`(${placeholders.join(', ')})`);
        }

        let query = `INSERT INTO ${this.q}${this.tableName}${this.q} (${columns}) VALUES ${valueSets.join(', ')}`;
        
        if (this.dbType === 'postgres') {
            query += ' RETURNING *';
        }

        const [results, metadata] = await this.sequelize.query(query, {
            replacements,
            type: this.sequelize.QueryTypes.INSERT
        });

        return { 
            success: true, 
            message: isBulk ? `${records.length} records inserted successfully.` : "Record inserted successfully.",
            insertedRows: results 
        };
    }

    /**
     * Update existing records
     * @param {Object} conditions - WHERE clause
     * @param {Object} data - Columns to update
     */
    async update(conditions, data) {
        if (!conditions || Object.keys(conditions).length === 0) {
            throw new Error("Update requires conditions to prevent bulk overwrite.");
        }

        const replacements = {};
        const setClause = Object.keys(data).map(k => `${this.q}${k}${this.q} = :data_${k}`).join(', ');
        for (const [k, v] of Object.entries(data)) {
            replacements[`data_${k}`] = v;
        }

        const conditionClauses = this._buildWhereClause(conditions, replacements, 'cond_');
        if (conditionClauses.length === 0) {
            throw new Error("Update conditions could not be generated. Aborting to prevent bulk update.");
        }
        
        const query = `UPDATE ${this.q}${this.tableName}${this.q} SET ${setClause} WHERE ${conditionClauses.join(' AND ')}`;

        const [results, metadata] = await this.sequelize.query(query, {
            replacements,
            type: this.sequelize.QueryTypes.UPDATE
        });

        return { success: true, message: "Record(s) updated successfully.", affectedRows: metadata };
    }

    /**
     * Delete records
     * @param {Object} conditions - WHERE clause
     */
    async delete(conditions) {
        if (!conditions || Object.keys(conditions).length === 0) {
            throw new Error("Delete requires conditions to prevent bulk delete.");
        }

        const replacements = {};
        const conditionClauses = this._buildWhereClause(conditions, replacements, 'del_');

        if (conditionClauses.length === 0) {
            throw new Error("Delete conditions could not be generated. Aborting to prevent bulk delete.");
        }

        const query = `DELETE FROM ${this.q}${this.tableName}${this.q} WHERE ${conditionClauses.join(' AND ')}`;

        const deleteResult = await this.sequelize.query(query, {
            replacements,
            type: this.sequelize.QueryTypes.DELETE
        });

        // Some dialects return [results, metadata], others return just metadata or undefined
        const affectedRows = Array.isArray(deleteResult) ? deleteResult[1] : deleteResult;

        return { success: true, message: "Record(s) deleted successfully.", affectedRows };
    }

    /**
     * Count records
     * @param {Object} conditions 
     */
    async count(conditions = {}) {
        let query = `SELECT COUNT(*) AS count FROM ${this.q}${this.tableName}${this.q}`;
        const replacements = {};

        const conditionClauses = this._buildWhereClause(conditions, replacements, 'c_');
        if (conditionClauses.length > 0) {
            query += ` WHERE ${conditionClauses.join(' AND ')}`;
        }

        const results = await this.sequelize.query(query, {
            replacements,
            type: this.sequelize.QueryTypes.SELECT,
            raw: true
        });

        return parseInt(results[0]?.count || 0, 10);
    }
}

module.exports = QueryBuilder;

File: pingdartdb-python/src/pingdartdb/query_builder.py
import math

class QueryBuilder:
    def __init__(self, table_name: str, connection, db_type: str):
        self.table_name = table_name
        self.connection = connection
        self.db_type = db_type
        self.q = '"' if db_type in ['postgresql', 'pgsql'] else '`'

    def _build_where_clause(self, conditions: dict):
        clauses = []
        replacements = []
        
        if not conditions or not isinstance(conditions, dict):
            return clauses, replacements

        for key, value in conditions.items():
            if isinstance(value, dict):
                # BETWEEN operator
                if 'min' in value and 'max' in value:
                    clauses.append(f"{self.q}{key}{self.q} BETWEEN %s AND %s")
                    replacements.extend([value['min'], value['max']])
                elif 'start' in value and 'end' in value:
                    clauses.append(f"{self.q}{key}{self.q} BETWEEN %s AND %s")
                    replacements.extend([value['start'], value['end']])
            elif isinstance(value, list):
                # IN clause
                if len(value) > 0:
                    placeholders = ', '.join(['%s'] * len(value))
                    clauses.append(f"{self.q}{key}{self.q} IN ({placeholders})")
                    replacements.extend(value)
                else:
                    clauses.append("1=0")
            elif isinstance(value, str):
                if value.startswith('!'):
                    clauses.append(f"{self.q}{key}{self.q} != %s")
                    replacements.append(value[1:])
                elif value.startswith('>=') or value.startswith('=>'):
                    clauses.append(f"{self.q}{key}{self.q} >= %s")
                    replacements.append(value[2:])
                elif value.startswith('<=') or value.startswith('=<'):
                    clauses.append(f"{self.q}{key}{self.q} <= %s")
                    replacements.append(value[2:])
                elif value.startswith('>'):
                    clauses.append(f"{self.q}{key}{self.q} > %s")
                    replacements.append(value[1:])
                elif value.startswith('<'):
                    clauses.append(f"{self.q}{key}{self.q} < %s")
                    replacements.append(value[1:])
                else:
                    clauses.append(f"{self.q}{key}{self.q} = %s")
                    replacements.append(value)
            else:
                clauses.append(f"{self.q}{key}{self.q} = %s")
                replacements.append(value)
                
        return clauses, replacements

    def _process_margedata_batch(self, rows: list, margedata_item: dict, search: dict):
        target_table = margedata_item.get('target_table')
        target_column = margedata_item.get('target_column')
        target_value = margedata_item.get('target_value')
        target_label = margedata_item.get('target_label')
        search_fields = margedata_item.get('search_fields', [])
        nested_margedata = margedata_item.get('margedata', [])
        range_data = margedata_item.get('range')

        if not all([target_table, target_column, target_value, target_label]):
            return

        parent_ids = list(set([row[target_value] for row in rows if target_value in row and row[target_value] is not None]))
        
        if not parent_ids:
            for row in rows:
                row[target_label] = []
            return

        conditions = []
        replacements = []

        placeholders = ', '.join(['%s'] * len(parent_ids))
        conditions.append(f"{self.q}{target_column}{self.q} IN ({placeholders})")
        replacements.extend(parent_ids)

        for field in search_fields:
            for search_key, full_column in field.items():
                parts = full_column.split('.')
                column = parts[1] if len(parts) > 1 else parts[0]
                if search and search_key in search and search[search_key]:
                    like_op = 'ILIKE' if self.db_type == 'postgresql' else 'LIKE'
                    conditions.append(f"{self.q}{column}{self.q} {like_op} %s")
                    replacements.append(f"%{search[search_key]}%")

        if isinstance(range_data, dict) and all(k in range_data for k in ['latitude', 'longitude', 'radius', 'target_latitude', 'target_longitude']):
            earth_radius = 6371
            conditions.append(f"""
                ({earth_radius} * ACOS(
                    COS(RADIANS(%s)) *
                    COS(RADIANS({self.q}{range_data['target_latitude']}{self.q})) *
                    COS(RADIANS({self.q}{range_data['target_longitude']}{self.q}) - RADIANS(%s)) +
                    SIN(RADIANS(%s)) *
                    SIN(RADIANS({self.q}{range_data['target_latitude']}{self.q}))
                )) <= %s
            """)
            replacements.extend([
                range_data['latitude'],
                range_data['longitude'],
                range_data['latitude'],
                range_data['radius']
            ])

        query = f"SELECT * FROM {self.q}{target_table}{self.q} WHERE {' AND '.join(conditions)}"
        
        try:
            with self.connection.cursor() as cursor:
                cursor.execute(query, replacements)
                all_matches = cursor.fetchall()
                
            grouped_matches = {}
            for match in all_matches:
                pid = match[target_column]
                if pid not in grouped_matches:
                    grouped_matches[pid] = []
                grouped_matches[pid].append(match)
                
            for row in rows:
                pid = row[target_value]
                row[target_label] = grouped_matches.get(pid, [])
                
            if nested_margedata and all_matches:
                for child_item in nested_margedata:
                    self._process_margedata_batch(all_matches, child_item, search)
                    
        except Exception as e:
            print(f"Error in batch margedata ({target_label}): {e}")
            for row in rows:
                row[target_label] = []

    def read(self, options: dict = None):
        if options is None:
            options = {}
            
        conditions = options.get('conditions', {})
        search = options.get('search', {})
        order_by = options.get('orderBy')
        pagination = options.get('pagination')
        range_data = options.get('range')
        margedata = options.get('margedata', [])

        query = f"SELECT * FROM {self.q}{self.table_name}{self.q} WHERE 1=1"
        replacements = []

        cond_clauses, cond_reps = self._build_where_clause(conditions)
        if cond_clauses:
            query += " AND " + " AND ".join(cond_clauses)
            replacements.extend(cond_reps)

        if isinstance(search, dict) and search:
            search_clauses = []
            for key, value in search.items():
                if value:
                    like_op = 'ILIKE' if self.db_type == 'postgresql' else 'LIKE'
                    search_clauses.append(f"{self.q}{key}{self.q} {like_op} %s")
                    replacements.append(f"%{value}%")
            if search_clauses:
                query += " AND (" + " OR ".join(search_clauses) + ")"

        if isinstance(range_data, dict) and all(k in range_data for k in ['latitude', 'longitude', 'radius', 'target_latitude', 'target_longitude']):
            earth_radius = 6371
            join_table = range_data.get('table_name')
            
            if not join_table or join_table == self.table_name:
                query += f""" AND (
                    {earth_radius} * ACOS(
                        COS(RADIANS(%s)) *
                        COS(RADIANS({self.q}{range_data['target_latitude']}{self.q})) *
                        COS(RADIANS({self.q}{range_data['target_longitude']}{self.q}) - RADIANS(%s)) +
                        SIN(RADIANS(%s)) *
                        SIN(RADIANS({self.q}{range_data['target_latitude']}{self.q}))
                    )
                ) <= %s"""
                replacements.extend([range_data['latitude'], range_data['longitude'], range_data['latitude'], range_data['radius']])
            else:
                alias = 'rt'
                join_col = range_data.get('join_column', 'user_id')
                main_col = range_data.get('main_join_column', 'id')
                query += f""" AND EXISTS (
                    SELECT 1 FROM {self.q}{join_table}{self.q} AS {alias}
                    WHERE {alias}.{self.q}{join_col}{self.q} = {self.q}{self.table_name}{self.q}.{self.q}{main_col}{self.q}
                    AND (
                        {earth_radius} * ACOS(
                            COS(RADIANS(%s)) *
                            COS(RADIANS({alias}.{self.q}{range_data['target_latitude']}{self.q})) *
                            COS(RADIANS({alias}.{self.q}{range_data['target_longitude']}{self.q}) - RADIANS(%s)) +
                            SIN(RADIANS(%s)) *
                            SIN(RADIANS({alias}.{self.q}{range_data['target_latitude']}{self.q}))
                        )
                    ) <= %s
                )"""
                replacements.extend([range_data['latitude'], range_data['longitude'], range_data['latitude'], range_data['radius']])

        if order_by:
            query += f" ORDER BY {order_by}"

        count_query = query.replace("SELECT *", "SELECT COUNT(*) as total", 1)
        
        with self.connection.cursor() as cursor:
            cursor.execute(count_query, replacements)
            total_count = cursor.fetchone()['total']

        total_pages = 0
        if isinstance(pagination, dict) and 'page' in pagination and 'limit' in pagination:
            page = max(1, int(pagination['page']))
            limit = int(pagination['limit'])
            offset = (page - 1) * limit
            query += f" LIMIT {limit} OFFSET {offset}"
            total_pages = math.ceil(total_count / limit)

        with self.connection.cursor() as cursor:
            cursor.execute(query, replacements)
            results = cursor.fetchall()

        if margedata and results:
            for md_item in margedata:
                self._process_margedata_batch(results, md_item, search)

        return {
            'success': True,
            'data': results,
            'totalCount': total_count,
            'totalPages': total_pages
        }

    def insert(self, arg1, arg2=None):
        data = None
        conditions = None

        if arg2 is not None:
            data = arg2
            if isinstance(arg1, dict) and 'conditions' in arg1:
                conditions = arg1['conditions']
        else:
            data = arg1

        if not data:
            raise Exception("Data object cannot be empty")

        is_bulk = isinstance(data, list)
        records = data if is_bulk else [data]

        if not records:
            raise Exception("Data array cannot be empty")

        if isinstance(conditions, dict) and conditions:
            if not is_bulk:
                cond_clauses, check_reps = self._build_where_clause(conditions)
                if cond_clauses:
                    check_query = f"SELECT COUNT(*) as total FROM {self.q}{self.table_name}{self.q} WHERE " + " AND ".join(cond_clauses)
                    with self.connection.cursor() as cursor:
                        cursor.execute(check_query, check_reps)
                        existing_count = cursor.fetchone()['total']
                        if existing_count > 0:
                            return {
                                'success': False,
                                'exists': True,
                                'message': 'Record already exists matching the given conditions. Insert skipped.'
                            }
            else:
                condition_keys = list(conditions.keys())
                to_insert = []
                skipped = []

                for record in records:
                    per_record_conditions = {}
                    for key in condition_keys:
                        per_record_conditions[key] = record.get(key, conditions[key])
                        
                    cond_clauses, check_reps = self._build_where_clause(per_record_conditions)
                    if cond_clauses:
                        check_query = f"SELECT COUNT(*) as total FROM {self.q}{self.table_name}{self.q} WHERE " + " AND ".join(cond_clauses)
                        with self.connection.cursor() as cursor:
                            cursor.execute(check_query, check_reps)
                            if cursor.fetchone()['total'] > 0:
                                skipped.append(record)
                            else:
                                to_insert.append(record)
                    else:
                        to_insert.append(record)

                if not to_insert:
                    return {
                        'success': False,
                        'exists': True,
                        'message': f"All {len(records)} records already exist. Insert skipped.",
                        'skipped': len(skipped)
                    }
                records = to_insert # Swap for actual execution
                
        keys = list(records[0].keys())
        columns = ", ".join([f"{self.q}{k}{self.q}" for k in keys])
        
        value_sets = []
        replacements = []
        
        for record in records:
            placeholders = ", ".join(["%s"] * len(keys))
            value_sets.append(f"({placeholders})")
            replacements.extend([record.get(k) for k in keys])

        query = f"INSERT INTO {self.q}{self.table_name}{self.q} ({columns}) VALUES " + ", ".join(value_sets)
        
        with self.connection.cursor() as cursor:
            cursor.execute(query, replacements)
            inserted_rows = cursor.rowcount

        msg = f"{len(records)} record(s) inserted. {len(records) - len(records)} duplicate(s) skipped." if is_bulk and conditions else \
              f"{len(records)} records inserted successfully." if is_bulk else "Record inserted successfully."

        res = {
            'success': True,
            'message': msg,
            'insertedRows': inserted_rows
        }
        
        if is_bulk and conditions:
            res['inserted'] = len(records)
            res['skipped'] = len(data) - len(records)
            
        return res

    def update(self, data: dict, conditions: dict):
        if not conditions:
            raise Exception("Update requires conditions to prevent bulk overwrite.")

        set_clauses = []
        replacements = []
        
        for k, v in data.items():
            set_clauses.append(f"{self.q}{k}{self.q} = %s")
            replacements.append(v)

        cond_clauses, cond_reps = self._build_where_clause(conditions)
        if not cond_clauses:
            raise Exception("Update conditions could not be generated. Aborting to prevent bulk update.")
            
        replacements.extend(cond_reps)

        query = f"UPDATE {self.q}{self.table_name}{self.q} SET {', '.join(set_clauses)} WHERE {' AND '.join(cond_clauses)}"
        
        with self.connection.cursor() as cursor:
            cursor.execute(query, replacements)
            affected = cursor.rowcount

        return {
            'success': True,
            'message': "Record(s) updated successfully.",
            'affectedRows': affected
        }

    def delete(self, conditions: dict):
        if not conditions:
            raise Exception("Delete requires conditions to prevent bulk delete.")

        cond_clauses, replacements = self._build_where_clause(conditions)
        if not cond_clauses:
            raise Exception("Delete conditions could not be generated. Aborting to prevent bulk delete.")

        query = f"DELETE FROM {self.q}{self.table_name}{self.q} WHERE {' AND '.join(cond_clauses)}"
        
        with self.connection.cursor() as cursor:
            cursor.execute(query, replacements)
            affected = cursor.rowcount

        return {
            'success': True,
            'message': "Record(s) deleted successfully.",
            'affectedRows': affected
        }

    def count(self, conditions: dict = None):
        if conditions is None:
            conditions = {}
            
        query = f"SELECT COUNT(*) as count FROM {self.q}{self.table_name}{self.q}"
        cond_clauses, replacements = self._build_where_clause(conditions)
        
        if cond_clauses:
            query += " WHERE " + " AND ".join(cond_clauses)
            
        with self.connection.cursor() as cursor:
            cursor.execute(query, replacements)
            return cursor.fetchone()['count']
